add docs about activities in Node (#473)
temporalio/documentation: docs/node/activities.md

# Activities

In Temporal, Activities are typically used to represent side effects, like making an HTTP request.

> Unlike [Workflows](/docs/node/logging#logging-from-workflows-and-activities), activities run in the standard Node.js environment, not an [isolate](https://www.npmjs.com/package/isolated-vm).

Reviewer: Is there a better page we can link for workflows? Maybe the one about determinism?

> So any code that needs to talk to the outside world, including importing modules, needs to be in an Activity.

Reviewer: Not true, you can import any module in WFs as long as it doesn't use the node or dom libs.

### Overview

Below is a simple Activity that accepts a string parameter, appends a word to it, and returns the result.

> The Temporal Node SDK looks for any `.js` files in the `src/activities` directory, and automatically registers any exported functions as Activities.

Reviewer: Not accurate, in the package initializer, it's `.ts` in `src/activites` and the compiled code in `lib/activities` (configurable).

Author: Worth it to always use `lib/activities` rather than `src/activities`? If you're not using package initializer, either seems to work because of what [`activitiesPath` resolves to](https://github.com/temporalio/sdk-node/blob/9176b27270f7da77972ceb18c9c624f12e11ee93/packages/worker/src/worker.ts#L329)

# Activities

> In Temporal, Activities are typically used to represent side effects, like making an HTTP request.

Reviewer 2: "side effect" is actually the name of an API used to do some "non-deterministic" logic inside the Workflow code (at least in some of the other languages). Using it here as a way to describe an Activity is not ideal I think.

```suggestion
In Temporal, Activities are typically used to interact with external resources, like making an HTTP request.
```

> Unlike [Workflows](/docs/node/determinism), activities run in the standard Node.js environment, not an [isolate](https://www.npmjs.com/package/isolated-vm).

Reviewer 2: ```suggestion
Unlike [Workflows](/docs/node/determinism), Activities execute in the standard Node.js environment, not an [isolate](https://www.npmjs.com/package/isolated-vm).
```

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

<Tabs
defaultValue="ts"
values={[
{label: 'TypeScript', value: 'ts'},
{label: 'JavaScript', value: 'js'},
]
}>

<TabItem value="ts">

```typescript
export async function greet(name: string): Promise<string> {
  return `Hello, ${name}!`;
}
```

</TabItem>
<TabItem value="js">

```javascript
"use strict";

async function greet(name) {
  return `Hello, ${name}!`;
}
exports.greet = greet;
```

</TabItem>
</Tabs>

Under the hood, the Temporal Node SDK uses [Webpack](https://webpack.js.org/) to bundle your Activities before running them.
This [may cause issues with certain npm modules](https://www.getrevue.co/profile/masteringjs/issues/why-i-m-not-using-webpack-for-lambda-functions-anymore-266010).

### Importing and Using

You can import the above `greet()` Activity in a Workflow as shown below, assuming that the `greet` function is in the `lib/activities/greeter.js` file.

```typescript
import {greet} from "@activities/greeter";
```

You can also use `require()` as shown below.

```javascript
const {greet} = require("@activities/greeter");
```

:::info

Currently, you need to use the `@activities` prefix.

> Do **not** import your Activity using the `'../activities/greeter'` path, otherwise your Activity will run in the same V8 isolate as your Workflow and your Activity will be subject to the same restrictions as your Workflow.

Reviewer 2: ```suggestion
Do **not** import your Activity using the `'../activities/greeter'` path, otherwise your Activity will execute in the same V8 isolate as your Workflow and your Activity will be subject to the same restrictions as your Workflow.
```

> Do **not** import your Activity using the `'../activities/greeter'` path, otherwise your Activity will run in the same V8 isolate as your Workflow and your Activity will be subject to the same restrictions as your Workflow.

Reviewer 3: ```suggestion
Do **not** import your Activity by using the `'../activities/greeter'` path; otherwise your Activity runs in the same V8 isolate as your Workflow, and your Activity is subject to the same restrictions as your Workflow.
```

Author: Thanks :+1:

# Activities

In Temporal, Activities are typically used to interact with external resources, like making an HTTP request.
Unlike [Workflows](/docs/node/determinism), Activities execute in the standard Node.js environment, not an [isolate](https://www.npmjs.com/package/isolated-vm).
So any code that needs to talk to the outside world needs to be in an Activity.

### Overview

Below is a simple Activity that accepts a string parameter, appends a word to it, and returns the result.
The Temporal Node SDK looks for any `.js` files in the `lib/activities` directory, and automatically registers any exported functions as Activities.

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

<Tabs
defaultValue="ts"
values={[
{label: 'TypeScript', value: 'ts'},
{label: 'JavaScript', value: 'js'},
]
}>

<TabItem value="ts">

```typescript
export async function greet(name: string): Promise<string> {
  return `Hello, ${name}!`;
}
```

</TabItem>
<TabItem value="js">

```javascript
"use strict";

async function greet(name) {
  return `Hello, ${name}!`;
}
exports.greet = greet;
```

</TabItem>
</Tabs>

> Under the hood, the Temporal Node SDK uses [Webpack](https://webpack.js.org/) to bundle your Activities before running them.

Reviewer: Activities don't get bundled, only Workflows.
